### ISSUE TITLE
How to use onnx classify model at C++ 

### ISSUE THREAD

**Issue by freedomdive** [ultralytics/ultralytics#13337]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I have C++ api for predictions using onnx-runtime-api.
Firstly I train models via yolo train, next exporting best.pt to *.onnx with opset=12.
Next I get vector<box> with labels.
It works fine with task object detection. Results are the same for yolo *.pt scripts and this api.

But when I try to use classification model - doing the same things - yolo classify predict works perfect, but *.onnx model writes "random" probabilities at output tensors.

I use image changes to float [0 - 1] values with image/255, using correct model sizes like set via train and so on according to instructions. RGB order is correct.

Changes are at output tensors:
At object detection trained model there are lots of data that I parse via nms logic and get correct labels.
At classification trained model there are [number of classes] float values. And they are not 1.0 0.0 0.0 like at yolo classify predict results info. (I read that they must be confidents of given classes via indexes order).

Are there anything else I should change using classification pretrained models and onnx-runtime-api?

Additional info:

`yolo classify predict model=best.pt source=image.png
Ultralytics YOLOv8.0.103  Python-3.9.13 torch-1.9.0+cu111 CUDA:0 (NVIDIA GeForce RTX 3080, 10240MiB)
YOLOv8n-cls summary (fused): 73 layers, 1457938 parameters, 0 gradients, 3.3 GFLOPs

image 1/1 [path]\image.png: 224x224 500023_0 1.00, 10004_0 0.00, 500023_1 0.00, 100004_0 0.00, 500097_0 0.00, 15.6ms
Speed: 0.0ms preprocess, 15.6ms inference, 0.0ms postprocess per image at shape (1, 3, 224, 224)`

500023_0 is correct class. 

Next I export best.pt to onnx via yolo export format=onnx model=best.pt opset=12

Load this model, setting 224x224 resize as needed.

Number of classes is 18

![image](https://github.com/ultralytics/ultralytics/assets/17836243/df2753b0-5191-4440-93d0-8b9549f62b75)

And recieve something differ than 1.0 0.0 0.0 and so on.
Best prediction around 0.3 and it is incorrect.

Firstly I tried to do the same with setting one label at object detection mode and all labels are correct with around 1.0 confidence.

**Comment by glenn-jocher** [comment]
@freedomdive hello! It looks like you've done a thorough job ensuring your setup is correct for using the ONNX model in a C++ environment. Here are a few suggestions that might help resolve the issue with the classification model outputs:

1. **Model Output Activation**: Ensure that the output layer of your classification model is using a softmax activation function. The YOLO classification models are trained with a softmax layer to output probabilities. If the exported ONNX model does not apply softmax, the outputs will be logits, which could explain the "random" probabilities you're observing.

2. **Post-Processing**: After obtaining the output from the ONNX model, apply a softmax function to convert these logits into probabilities. This can be done easily in C++ using a standard softmax implementation.

3. **Check ONNX Export**: Verify that the ONNX model is exported correctly with all necessary layers and activations. Sometimes, during the conversion, certain optimizations or layer transformations might alter the expected behavior of the model.

4. **Model Input Normalization**: Double-check that the input normalization (scaling of pixel values, color order) exactly matches what was used during training. Any discrepancy here can lead to poor model performance.

5. **Debugging Tip**: Compare the outputs directly from the PyTorch model and the ONNX model using the same input image to ensure that the conversion has no issues.

If these steps do not resolve the issue, it might be helpful to look into the specific layers and operations in the ONNX model to ensure they match the original PyTorch model. Sometimes, visualizing the model architecture can help spot discrepancies.

I hope this helps! Let us know if you have further questions.

**Comment by glenn-jocher** [comment]
@freedomdive, great to hear that the ONNX model works correctly with the YOLO predict script! 🎉

Regarding the discrepancies you're seeing with the C++ ONNX Runtime API, it's good that you're planning to dive deeper into the Python scripts. Understanding the preprocessing and post-processing steps in Python could indeed reveal critical differences.

A few additional suggestions:
- **Ensure Consistency in Preprocessing**: Double-check that the image preprocessing in C++ matches exactly with Python. This includes image resizing, normalization, and color channel ordering.
- **Debugging Output**: Consider logging the raw outputs directly before the softmax application in both Python and C++ to compare.
- **ONNX Model Inspection**: Use tools like Netron to inspect the ONNX model to ensure that all layers and operations are as expected.

Keep us updated on your progress, and don't hesitate to reach out if you hit another snag!

**Comment by freedomdive** [comment]
@glenn-jocher, I partially solved the problem.

I rescaled all images the same way I rescale them in C++ before converting to float32 (to prevent Python from rescaling during training) and trained the model on this data.

Now the model gives correct confidence scores:

![image](https://github.com/ultralytics/ultralytics/assets/17836243/7e579af1-63ef-4a05-b224-e9221578326b)

Result is correct.

Interesting, why object detection likes my C++ rescaling, and classification does not.

Here’s what I do during rescaling:
Custom bilinear affine transformation and padding with color 114.

**Comment by glenn-jocher** [comment]
@freedomdive, fantastic progress! 🌟 It's intriguing to see how the rescaling approach impacted the classification results differently from object detection. This could be due to how the two tasks process spatial information and color consistency. Classification might be more sensitive to subtle variations in input preprocessing.

Your approach to standardize image preprocessing across both training and inference is spot-on. It ensures consistency, which is crucial for model performance. The custom bilinear affine transformation and specific padding color you're using in C++ might introduce slight differences that affect classification more than detection.

Keep up the great work, and thanks for sharing your findings! It adds valuable insights into the nuances of model training and deployment. If you have further questions or new updates, feel free to share!
